{"blogid": "125649451", "writerAge": "码龄2年", "writerBlogNum": "37", "writerCollect": "620", "writerComment": "121", "writerFan": "48", "writerGrade": "3级", "writerIntegral": "743", "writerName": "蓝胖胖▸", "writerProfileAdress": "writer_image\\profile_125649451.jpg", "writerRankTotal": "25272", "writerRankWeekly": "31275", "writerThumb": "228", "writerVisitNum": "80913", "blog_read_count": "10190", "blog_time": "已于 2022-07-08 17:37:12 修改", "blog_title": "一步真实解决RuntimeError: result type Float can‘t be cast to the desired output type long int", "content": "<div class=\"article_content clearfix\" id=\"article_content\">\n<link href=\"style.css\" rel=\"stylesheet\"/>\n<div class=\"htmledit_views\" id=\"content_views\">\n<h1><a id=\"_0\"></a>问题描述：</h1>\n<p>在使用YOLO框架训练自己的数据集时候，你是不是出现了这个问题？</p>\n<p><img alt=\"\" height=\"704\" src=\"image\\ac7eea97a0d544fe9a23b9d4126ea678.png\" width=\"1200\"/></p>\n<blockquote>\n<p> RuntimeError: result type Float can't be cast to the desired output type long int<code class=\"language-c\"> </code></p>\n</blockquote>\n<p>明明自己【data】下的yaml文件，【models】下网络结构等文件设置与修改都是按照教程来的</p>\n<p>可还是出现了问题？这是怎么回事？</p>\n<p>起初我也遇到了这个问题，但是经过思考，终于找到了问题的根源：</p>\n<blockquote>\n<p>官网的yolov5-master版本可以正常运行，但是yolov5-5.0/yolov5-6.1等版本就是不可以运行</p>\n<p></p>\n<p>这是因为yolov5-master版本和yolov5-5.0/yolov5-6.1等版本下的【utils】中的【loss.py】文件是不一样的，大概是yolov5-5.0/yolov5-6.1等版本在更新版本的时候出了问题</p>\n</blockquote>\n<p>不过，我已经总结出了解决这个问题办法，详细看下一章节内容吧~~~</p>\n<hr/>\n<h1><a id=\"_23\"></a>问题解决：</h1>\n<blockquote>\n<p><strong>修改【utils】中的【loss.py】里面的两处内容</strong></p>\n</blockquote>\n<p> 1.打开你的【utils】文件下的【loss.py】</p>\n<p>2.按【Ctrl】+【F】打开搜索功能，输入【for i in range(self.nl)】找到下面的一行内容：</p>\n<p><img alt=\"\" height=\"128\" src=\"image\\ac1badafe55845fcb23996aa261b4cbb.png\" width=\"754\"/></p>\n<p> （上面的代码在【loss.py】的后半部分）</p>\n<p>将下面的代码替换掉上图中的红圈部分：</p>\n<pre><code> anchors, shape = self.anchors[i], p[i].shape </code></pre>\n<p> 替换之后的效果是这样的：</p>\n<p><img alt=\"\" height=\"132\" src=\"image\\ad7e4e9e01d443168a25dd8bf244e6a5.png\" width=\"799\"/></p>\n<p>3.按【Ctrl】+【F】打开搜索功能，输入【indices.append】找到下面的一行内容： </p>\n<p><img alt=\"\" height=\"170\" src=\"image\\3b9b9a2396ce44cabaf6314464fd0d5a.png\" width=\"1028\"/></p>\n<p>  （上面的代码在【loss.py】的最后部分，具体位置在上一处搜索位置的下面10行左右）</p>\n<p> 将下面的代码替换掉上图中的红圈部分：</p>\n<pre><code>indices.append((b, a, gj.clamp_(0, shape[2] - 1), gi.clamp_(0, shape[3] - 1)))  # image, anchor, grid</code></pre>\n<p>替换之后的效果是这样的：<img alt=\"\" height=\"191\" src=\"image\\100a40a78d5f46f1935a78fbaec8d38c.png\" width=\"1113\"/></p>\n<p>4.保存刚才的两个替换操作 </p>\n<p>5.再次运行【train..py】文件，问题解决！！！代码跑通！！！<img alt=\"\" src=\"image\\e6451c065eb74b10a0c1db00983e4b05.jpeg\"/></p>\n<p><strong>6.如果你还是跑不通，可以看最后的小提示，可能会有你的问题！ </strong></p>\n<hr/>\n<h2>【更新】小提示</h2>\n<p>如果你按照我的方法改了指定的代码，还出现了下面的问题？</p>\n<p><img alt=\"\" height=\"259\" src=\"image\\e02fba0da26b45659ac46ae2c6d3acbc.png\" width=\"985\"/></p>\n<p>那么你是在修改下面地方时，没有将【anchors】和下面的【gain】对齐！！！</p>\n<p><img alt=\"\" height=\"117\" src=\"image\\c976d5d8e4c34ee4b2d70bf43d962fdb.png\" width=\"499\"/></p>\n<p> 打开【loss,.py】对齐下代码就可以了！完美解决小小粗心bug~</p>\n<hr/>\n<h3> 到这，就已经解决了问题。欢迎关注蓝胖胖，咱们一起解决BUG~~~</h3>\n</div>\n</div>"}